heartbeat/monitors/monitor.go

```diff
@@ -0,0 +1,120 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package monitors
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/heartbeat/scheduler"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Monitor represents a configured recurring monitoring task loaded from a config file. Starting it
+// will cause it to run with the given scheduler until Stop() is called.
+type Monitor struct {
+	name       string
+	config     *common.Config
+	registrar  *pluginsReg
+	uniqueName string
+	scheduler  *scheduler.Scheduler
+	jobTasks   map[string]*monitorTask
+	enabled    bool
+
+	pipeline beat.Pipeline
+}
+
+func checkMonitorConfig(config *common.Config, registrar *pluginsReg) error {
+	_, err := newMonitor(config, registrar, nil, nil)
+	return err
+}
+
+func newMonitor(
+	config *common.Config,
+	registrar *pluginsReg,
+	pipeline beat.Pipeline,
+	scheduler *scheduler.Scheduler,
+) (*Monitor, error) {
+	// Extract just the Type and Enabled fields from the config
+	// We'll parse things more precisely later once we know what exact type of
+	// monitor we have
+	mpi, err := pluginInfo(config)
+	if err != nil {
+		return nil, err
+	}
+
+	monitorPlugin, found := registrar.get(mpi.Type)
+	if !found {
+		return nil, fmt.Errorf("monitor type %v does not exist", mpi.Type)
+	}
+
+	m := &Monitor{
+		name:      monitorPlugin.name,
+		scheduler: scheduler,
+		jobTasks:  map[string]*monitorTask{},
+		pipeline:  pipeline,
+	}
+
+	jobs, err := monitorPlugin.create(config)
+	if err != nil {
+		return nil, fmt.Errorf("job err %v", err)
+	}
+
+	m.makeTasks(config, jobs)
+
+	return m, nil
+}
+
+func (m *Monitor) makeTasks(config *common.Config, jobs []Job) error {
+	mtConf := monitorTaskConfig{}
+	if err := config.Unpack(&mtConf); err != nil {
+		return errors.Wrap(err, "unexpected error while configuring heartbeat job")
```

Comment:
It's more like an 'invalid config' error. Unpacking/parsing the config into mtConf failed due to mtConf and the configuration object passed being incompatible.